**<p style = "text-align: center">TRƯỜNG ĐẠI HỌC KHOA HỌC TỰ NHIÊN - ĐẠI HỌC QUỐC GIA TPHCM</p>**

# **<p style = "text-align: center">NHẬP MÔN KHOA HỌC DỮ LIỆU</p>**
## **<p style = "text-align: center">Đồ án Cuối kì</p>**
### <p style = "text-align: center"> <b>Đề tài:</b> Phân tích, khám phá dữ liệu về sách trên web thương mại điện tử (tiki)</p>
**Sinh viên thực hiện**
| Mã số sinh viên | Họ và tên |
| --- | --- |
| 20120040 | Nguyễn Quang Gia Bảo |
| 20120136 | Huỳnh Tuấn Nam |
| 20120158 | Trần Hoàng Anh Phi |

---

# **1. Thông tin về dữ liệu**<br>
Dữ liệu được sử dụng trong bài được lấy từ trang: <a href = "https://tiki.vn/sach-truyen-tieng-viet/c316"> https://tiki.vn/sach-truyen-tieng-viet/c316</a><br>

Vì website <a href = 'https://tiki.vn/'> https://tiki.vn/ </a> có khá là nhiều mặt hàng nên nhóm chỉ chọn ra danh mục <font color=red>sách tiếng Việt</font> để crawl và thực hiện đồ án này.


# **2. Khám phá dữ liệu**<br>

#### Import các package cần thiết

In [200]:
import pandas as pd


#### Đọc dữ liệu đã thu thập từ file csv 

In [201]:
df = pd.read_csv('full_data/data.csv')
print(df.shape)
display(df.head())

C:\Users\Admin\AppData\Local\Temp\ipykernel_18800\906639567.py:1: DtypeWarning: Columns (22,27,29,30,31,32,33,34,35,36,37,38,39,40,41,43,45,46,47,48,49,50,51,52,53,54,56,57,58,59,60,61,62,63,64,65,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('full_data/data.csv')


(96647, 84)


Unnamed: 0         id  master_id            sku  \
0           0  204317934  204317934  9972713357587   
1           1  205279963  205279963  7618382614299   
2           2  193209826  193209826  5001141446127   
3           3  205288789  205288789  9035774965791   
4           4  204649516  204649516  8708257676679   

                                                name  \
0     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
1     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
2  Con Đường Tơ Lụa - Vạn Dặm Xa Từ Pakistan Tới ...   
3     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
4  Sách - 1111 - NHẬT KÝ SÁU VẠN DẶM TRÊN YÊN XE ...   

                                           short_url book_cover   price  \
0  https://tiki.vn/product-p204317934.html?spid=2...        NaN  262000   
1  https://tiki.vn/product-p205279963.html?spid=2...    Bìa mềm  256750   
2  https://tiki.vn/product-p193209826.html?spid=1...    Bìa mềm   95100   
3  https://tiki.vn/product-p205288789.html?spid=2...    Bìa mềm  292500   
4  https://tiki.vn/product-p204649516.html?spid=2...        NaN  292500   

   original_price  discount_rate  ...  kich_thuoc  tai_trong device_brand  \
0          325000             19  ...         NaN        NaN          NaN   
1          325000             21  ...         NaN        NaN          NaN   
2          129000             26  ...         NaN        NaN          NaN   
3          325000             10  ...         NaN        NaN          NaN   
4          325000             10  ...         NaN        NaN          NaN   

  do_nhay  tan_so huong_dan_bao_quan  dac_diem_noi_bat dieu_kien_su_dung  \
0     NaN     NaN                NaN               NaN               NaN   
1     NaN     NaN                NaN               NaN               NaN   
2     NaN     NaN                NaN               NaN               NaN   
3     NaN     NaN                NaN               NaN               NaN   
4     NaN     NaN                NaN               NaN               NaN   

  dia_chi_su_dung bluetooth  
0             NaN       NaN  
1             NaN       NaN  
2             NaN       NaN  
3             NaN       NaN  
4             NaN       NaN  

[5 rows x 84 columns]

***

#### **Cleaning data**

Drop cột 'Unnamed: 0'

In [202]:
df = df.drop(columns='Unnamed: 0')
df.head()

id  master_id            sku  \
0  204317934  204317934  9972713357587   
1  205279963  205279963  7618382614299   
2  193209826  193209826  5001141446127   
3  205288789  205288789  9035774965791   
4  204649516  204649516  8708257676679   

                                                name  \
0     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
1     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
2  Con Đường Tơ Lụa - Vạn Dặm Xa Từ Pakistan Tới ...   
3     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
4  Sách - 1111 - NHẬT KÝ SÁU VẠN DẶM TRÊN YÊN XE ...   

                                           short_url book_cover   price  \
0  https://tiki.vn/product-p204317934.html?spid=2...        NaN  262000   
1  https://tiki.vn/product-p205279963.html?spid=2...    Bìa mềm  256750   
2  https://tiki.vn/product-p193209826.html?spid=1...    Bìa mềm   95100   
3  https://tiki.vn/product-p205288789.html?spid=2...    Bìa mềm  292500   
4  https://tiki.vn/product-p204649516.html?spid=2...        NaN  292500   

   original_price  discount_rate  rating_average  ...  kich_thuoc tai_trong  \
0          325000             19             5.0  ...         NaN       NaN   
1          325000             21             0.0  ...         NaN       NaN   
2          129000             26             4.8  ...         NaN       NaN   
3          325000             10             0.0  ...         NaN       NaN   
4          325000             10             0.0  ...         NaN       NaN   

  device_brand  do_nhay tan_so  huong_dan_bao_quan dac_diem_noi_bat  \
0          NaN      NaN    NaN                 NaN              NaN   
1          NaN      NaN    NaN                 NaN              NaN   
2          NaN      NaN    NaN                 NaN              NaN   
3          NaN      NaN    NaN                 NaN              NaN   
4          NaN      NaN    NaN                 NaN              NaN   

  dieu_kien_su_dung dia_chi_su_dung bluetooth  
0               NaN             NaN       NaN  
1               NaN             NaN       NaN  
2               NaN             NaN       NaN  
3               NaN             NaN       NaN  
4               NaN             NaN       NaN  

[5 rows x 83 columns]

Xem thử dữ liệu có bị trùng hay không?

In [203]:
df.duplicated().sum()

8294

Vậy là dữ liệu có 8294 dòng trùng nhau, ta sẽ tiến hành xoá những dòng này

In [204]:
df = df.drop_duplicates(ignore_index=True)
df.shape

(88353, 83)

Vì khi thu thập dữ liệu, có những sản phẩm khác loại bị tiki xếp vào chung danh mục nên ta thấy những columns lạ bị thêm vào, ta phải tiến hành xoá những columns này và những sản phẩm <font color =red>lạc loài </font>

In [205]:
df.columns

Index(['id', 'master_id', 'sku', 'name', 'short_url', 'book_cover', 'price',
       'original_price', 'discount_rate', 'rating_average', 'review_count',
       'inventory_type', 'productset_group_name', 'day_ago_created',
       'categories', 'all_time_quantity_sold', 'authors', 'publisher_vn',
       'publication_date', 'dimensions', 'manufacturer', 'number_of_page',
       'dich_gia', 'edition', 'brand', 'brand_country', 'item_model_number',
       'origin', 'product_weight', 'battery_capacity', 'camera', 'chong_nuoc',
       'loai_day', 'san_pham', 'storage', 'battery_life', 'charge_time',
       'do_chiu_nuoc', 'included_accessories', 'material', 'luu_y',
       'minimum_inbound_policy_days', 'expiry_time', 'shelf_life_days',
       'chat_lieu', 'audio_power_output', 'audio_technology',
       'dieu_khien_bang_giong_noi', 'dieu_khien_tivi', 'digital_tivi_dvb_t2',
       'display_type', 'frequency', 'hdh_giao_dien',
       'image_processing_technology', 'network_hdmi', 'network_inte

In [206]:
df[~df['material'].isna()]

id  master_id            sku  \
13233  144356210  144356210  5525114755464   
14886  207035527  207035527  6816910771280   
15160  176689612  176689612  9588934896842   
25077  207631100  207631100  6801021851103   
49830  196565537  196565537  5928507424911   
63694  202432845  202432845  8183965466103   
76899  175150080  175150080  9992881446659   

                                                    name  \
13233  Túi đeo chạy bộ, túi đeo hông, đeo bụng dành c...   
14886  Áo Hoodie Nỉ Ngoại Nữ Thêu Chữ Loang Màu Hồng ...   
15160  Vỏ gối trắng trơn-vỏ gối cao cấp 100% Cotton-t...   
25077     Áo Khoác Nam Nữ Xanh Rêu Áo Khoác Dù Dáng Rộng   
49830  Rèm Tự Dính-Không Cần Khoan Vô Cùng Tiện Lợi, ...   
63694  BBSTORE'S Đầm Xòe Dự Tiệc Tay Phồng Thời Trang...   
76899  Bộ Bàn Ghế Gỗ Me Tây Nguyên Khối, Bàn Ghế Ăn G...   

                                               short_url book_cover     price  \
13233  https://tiki.vn/product-p144356210.html?spid=1...        NaN    112000   
14886  https://tiki.vn/product-p207035527.html?spid=2...        NaN    109000   
15160  https://tiki.vn/product-p176689612.html?spid=1...        NaN     63750   
25077  https://tiki.vn/product-p207631100.html?spid=2...        NaN    130000   
49830  https://tiki.vn/product-p196565537.html?spid=1...        NaN     52500   
63694  https://tiki.vn/product-p202432845.html?spid=2...        NaN    219000   
76899  https://tiki.vn/product-p175150080.html?spid=1...        NaN  20400000   

       original_price  discount_rate  rating_average  ...  \
13233          139000             19             4.8  ...   
14886          109000              0             0.0  ...   
15160           63750              0             0.0  ...   
25077          130000              0             0.0  ...   
49830          105000             50             0.0  ...   
63694          219000              0             0.0  ...   
76899        20400000              0             0.0  ...   

                 kich_thuoc tai_trong device_brand  do_nhay tan_so  \
13233                   NaN       NaN          NaN      NaN    NaN   
14886                   NaN       NaN          NaN      NaN    NaN   
15160                   NaN       NaN          NaN      NaN    NaN   
25077                   NaN       NaN          NaN      NaN    NaN   
49830           30 x 20 x 5       5kg          NaN      NaN    NaN   
63694                   NaN       NaN          NaN      NaN    NaN   
76899  3,39 x (70 - 75) x 5       NaN          NaN      NaN    NaN   

       huong_dan_bao_quan dac_diem_noi_bat dieu_kien_su_dung dia_chi_su_dung  \
13233                 NaN              NaN               NaN             NaN   
14886                 NaN              NaN               NaN             NaN   
15160                 NaN              NaN               NaN             NaN   
25077                 NaN              NaN               NaN             NaN   
49830                 NaN              NaN               NaN             NaN   
63694                 NaN              NaN               NaN             NaN   
76899                 NaN              NaN               NaN             NaN   

      bluetooth  
13233       NaN  
14886       NaN  
15160       NaN  
25077       NaN  
49830       NaN  
63694       NaN  
76899       NaN  

[7 rows x 83 columns]

Thử lấy các sản phẩm thuộc ngoài những `categories` của danh mục sách (mà ta đã thu thập + 3 danh mục mới)

In [207]:
feartures = ['id', 'master_id', 'sku', 'name', 'short_url', 'book_cover', 'price',
       'original_price', 'discount_rate', 'rating_average', 'review_count',
       'inventory_type', 'productset_group_name', 'day_ago_created',
       'categories', 'all_time_quantity_sold', 'authors', 'publisher_vn',
       'publication_date', 'dimensions', 'manufacturer', 'number_of_page',
       'dich_gia', 'edition', 'luu_y']
import numpy as np
unvalid_cols = list(set(df.columns) - set(feartures))
valid_categories = pd.read_csv('id_data/categories_id.csv')['categories'] 
valid_categories = '|'.join(list(valid_categories) + ['Root', 'Sách tiếng Việt', 'Nhà Sách Tiki'])
# df[df[unvalid_cols].notna().any(axis=1)].to_csv('not_books.csv', encoding='utf-8-sig')
# df[df['publisher_vn'].isna()].to_csv('not_publiser.csv', encoding='utf-8-sig')
# df[df['categories'].str.contains(valid_categories, na=False)]
df[~df['categories'].str.contains(valid_categories, na=False)]['categories'].value_counts()

Máy Ảnh - Máy Quay Phim                4
Túi và Hộp Đựng Máy Ảnh                3
Thiết Bị Số - Phụ Kiện Số              3
Tai Nghe True Wireless                 2
Thiết Bị Phòng Tối                     1
Thực phẩm bổ dưỡng                     1
Phụ kiện ô tô                          1
Bánh kẹo                               1
Móng giả                               1
Bộ sản phẩm vệ sinh nhà cửa            1
Làm Đẹp - Sức Khỏe                     1
Thực phẩm khô khác                     1
Loa Vi Tính                            1
Dây Đeo Máy Ảnh, Máy Quay, Ống Nhòm    1
Bình, ly uống trà và phụ kiện          1
Nước Ngọt                              1
Bàn ghế làm việc                       1
Nón bơi                                1
Phụ Kiện Máy Quay Khác                 1
Bao Da - Ốp Lưng iPad                  1
Bình đun siêu tốc                      1
Các loại gia vị dạng nước khác         1
Voucher - Dịch vụ                      1
Mystery, Thriller & Suspense           1
Thời Trang      

Và ta chỉ lấy những giá trị hợp lệ 

In [214]:
df = df[df['categories'].str.contains(valid_categories, na=False)]
# df.shape
df.shape

(87955, 83)

Có thể vẫn còn, ta sẽ xem những features bị <font color=red>lạc loài</font> (khác giá trị nan):

In [215]:
df[df[unvalid_cols].notna().any(axis=1)]

id  master_id            sku  \
14886  207035527  207035527  6816910771280   
15160  176689612  176689612  9588934896842   
15706   56784436   56784436  5214537897810   
16524   14713717   14713717  7685692038611   
16529   13119470   13119470  3574385530362   
25077  207631100  207631100  6801021851103   
29373  144180250  144180250  9731087042011   
31813  207614864  207614864  8861427856035   
32003  122347726  122347726  3481577897926   
32708  199844015  199844015  9895292024848   
32730  198573360  198573360  9707792175271   
35744  184857872  184857872  3476634950928   
37045  123237970  123237970  1945235425131   
43624  175782948  175782948  9878301461407   
61230  141525907  141525907  7436689667775   
61778   21563234   21563234  8054277093458   
62538  162125536  162125536  3843063567863   
63694  202432845  202432845  8183965466103   
75711   57042352   57042352  7063823984615   
77878   16375132   16375132  1830509343082   

                                                    name  \
14886  Áo Hoodie Nỉ Ngoại Nữ Thêu Chữ Loang Màu Hồng ...   
15160  Vỏ gối trắng trơn-vỏ gối cao cấp 100% Cotton-t...   
15706  Aquarium Fish Tank Air Pump Flow Tube Pipe Lin...   
16524                              Bộ lomo card GOT7 mới   
16529       Bộ postcard BTS Caramel 204 pcs mẫu mới nhất   
25077     Áo Khoác Nam Nữ Xanh Rêu Áo Khoác Dù Dáng Rộng   
29373  Dép lào xốp dép xỏ ngón dép tông lào siêu nhẹ ...   
31813                 Sweater Tay Dài Logo Tim Form Rộng   
32003                ĐÈN ÂM TRẦN( phi 90 ánh sáng 3 màu)   
32708  Tủ quần áo gỗ sồi 3 buồng (miễn phí vận chuyển...   
32730  Combo 2 Dây Buộc Tóc Lò Xo, Set 3 Thun Cột Tóc...   
35744  Bộ quần áo ngắn bé gái Cam trơn rũ cotton lụa ...   
37045  Kim Loại Kính Mát Gothic Phong Cách Khoa Học V...   
43624  Lưới Che Nắng Thái Lan Khổ Rộng 4m Hàng Loại 1...   
61230     Bồn cầu di động cho người già,trẻ nhỏ,người ốm   
61778  Ốp lưng nhựa dẻo dành cho Samsung J7 2016 in h...   
62538                                   Chân váy công sở   
63694  BBSTORE'S Đầm Xòe Dự Tiệc Tay Phồng Thời Trang...   
75711  Tranh treo tường phòng khách, phòng ngủ -Tranh...   
77878    Nước hoa nam Terre D'Evasion - Inessance 100ml   

                                               short_url book_cover    price  \
14886  https://tiki.vn/product-p207035527.html?spid=2...        NaN   109000   
15160  https://tiki.vn/product-p176689612.html?spid=1...        NaN    63750   
15706  https://tiki.vn/product-p56784436.html?spid=56...        NaN   147000   
16524  https://tiki.vn/product-p14713717.html?spid=14...        NaN    10000   
16529  https://tiki.vn/product-p13119470.html?spid=13...        NaN    36000   
25077  https://tiki.vn/product-p207631100.html?spid=2...        NaN   130000   
29373  https://tiki.vn/product-p144180250.html?spid=1...        NaN     9700   
31813  https://tiki.vn/product-p207614864.html?spid=2...        NaN   149000   
32003  https://tiki.vn/product-p122347726.html?spid=1...        NaN    38220   
32708  https://tiki.vn/product-p199844015.html?spid=1...        NaN  5600000   
32730  https://tiki.vn/product-p198573360.html?spid=1...        NaN    24000   
35744  https://tiki.vn/product-p184857872.html?spid=1...        NaN   165000   
37045  https://tiki.vn/product-p123237970.html?spid=1...        NaN   505701   
43624  https://tiki.vn/product-p175782948.html?spid=1...        NaN   336000   
61230  https://tiki.vn/product-p141525907.html?spid=1...        NaN   276000   
61778  https://tiki.vn/product-p21563234.html?spid=21...        NaN    49000   
62538  https://tiki.vn/product-p162125536.html?spid=1...        NaN   184800   
63694  https://tiki.vn/product-p202432845.html?spid=2...        NaN   219000   
75711  https://tiki.vn/product-p57042352.html?spid=17...        NaN   230000   
77878  https://tiki.vn/product-p16375132.html?spid=16...        NaN   345000   

       original_price  discount_rate  rating_average  ...  kich_thuoc  \
14886          109000           

Tiếp tục bỏ đi những sản phẩm này.

In [218]:
df = df.drop(df[unvalid_cols].notna().any(axis=1).index)
df[df[unvalid_cols].notna().any(axis=1)]

Empty DataFrame
Columns: [id, master_id, sku, name, short_url, book_cover, price, original_price, discount_rate, rating_average, review_count, inventory_type, productset_group_name, day_ago_created, categories, all_time_quantity_sold, authors, publisher_vn, publication_date, dimensions, manufacturer, number_of_page, dich_gia, edition, brand, brand_country, item_model_number, origin, product_weight, battery_capacity, camera, chong_nuoc, loai_day, san_pham, storage, battery_life, charge_time, do_chiu_nuoc, included_accessories, material, luu_y, minimum_inbound_policy_days, expiry_time, shelf_life_days, chat_lieu, audio_power_output, audio_technology, dieu_khien_bang_giong_noi, dieu_khien_tivi, digital_tivi_dvb_t2, display_type, frequency, hdh_giao_dien, image_processing_technology, network_hdmi, network_internet, network_wifi, remote_thong_minh, resolution, screen_mirroring, screen_size, size_without_stand_table_top, size_with_stand_table_top, tivi_type, ung_dung, usb, weight_without_stand, weight_with_stand, size, loai_pin, thoi_gian_su_dung, capacity, huong_dan_su_dung, kich_thuoc, tai_trong, device_brand, do_nhay, tan_so, huong_dan_bao_quan, dac_diem_noi_bat, dieu_kien_su_dung, dia_chi_su_dung, bluetooth]
Index: []

[0 rows x 83 columns]

Có thể vẫn còn những sản phẩm <font color=red>lạc loài</font> khác, tuy nhiên khó có thể nhận biết và loại bỏ được.